src/game/client/w3dsnow.cpp

```diff
@@ -0,0 +1,732 @@
+    m_unk6(0),
+    m_unk7(0),
+    m_currentParticleCount(0),
+    m_unk9(0)
+{
+}
+
+W3DSnowManager::~W3DSnowManager()
+{
+    Release_Resources();
+}
+
+void W3DSnowManager::Init()
+{
+    SnowManager::Init();
+    Re_Acquire_Resources();
+}
+
+void W3DSnowManager::Release_Resources()
+{
+    Ref_Ptr_Release(m_texture);
+
+#ifdef BUILD_WITH_D3D8
+    if (m_vertexBuffer) {
+        m_vertexBuffer->Release();
+    }
+    m_vertexBuffer = nullptr;
+#endif
+
+    Ref_Ptr_Release(m_indexBuffer);
+}
+
+bool W3DSnowManager::Re_Acquire_Resources()
+{
+#ifdef BUILD_WITH_D3D8
+    Release_Resources();
+
+    if (!g_theWeatherSetting->m_snowEnabled) {
+        return true;
+    }
+
+    if (g_theWeatherSetting->m_snowPointSprites && DX8Wrapper::Get_Current_Caps()->Supports_Large_Point_Size()) {
+        IDirect3DDevice8 *device = DX8Wrapper::Get_D3D_Device8();
+        captainslog_dbgassert(device != nullptr, "Trying to ReAquireResources on W3DSnowManager without device");
+
+        if (!m_vertexBuffer
+            && device->CreateVertexBuffer(VERTEX_COUNT * sizeof(Vector3),
+                D3DUSAGE_DYNAMIC | D3DUSAGE_POINTS | D3DUSAGE_WRITEONLY,
+                DX8_FVF_XYZ,
+                D3DPOOL_DEFAULT,
+                &m_vertexBuffer)) {
+            return false;
+        }
+    } else {
+        m_indexBuffer = new DX8IndexBufferClass(MAX_PARTICLE_COUNT * 6, DX8IndexBufferClass::USAGE_DEFAULT);
+
+        {
+            IndexBufferClass::WriteLockClass lock(m_indexBuffer, 0);
+
+            unsigned short *indices = lock.Get_Index_Array();
+            unsigned short vert = 0;
+
+            for (int i = 0; i < MAX_PARTICLE_COUNT; ++i) {
+                indices[0] = vert + 3;
+                indices[1] = vert + 0;
+                indices[2] = vert + 2;
+                indices[3] = vert + 2;
+                indices[4] = vert + 0;
+                indices[5] = vert + 1;
+                vert += 4;
+                indices += 6;
+            }
+        }
+    }
+
+    m_texture = W3DAssetManager::Get_Instance()->Get_Texture(g_theWeatherSetting->m_snowTexture.Str());
+    m_vertexBufferOffset = VERTEX_COUNT;
+    m_numVertices = VERTEX_COUNT;
+    m_maxParticleCount = MAX_PARTICLE_COUNT;
+
+    return true;
+#else
+    return false;
+#endif
+}
+
+inline unsigned long F2DW(float f)
+{
+    return *((unsigned *)&f);
+}
+
+void W3DSnowManager::Render(RenderInfoClass &rinfo)
+{
+#ifdef BUILD_WITH_D3D8
+    if (g_theWeatherSetting->m_snowEnabled && m_isVisible) {
+
+        bool use_point_sprites =
+            g_theWeatherSetting->m_snowPointSprites && DX8Wrapper::Get_Current_Caps()->Supports_Large_Point_Size();
+
+        Coord3D cam_coord = g_theTacticalView->Get_3D_Camera_Position();
+
+        Vector3 cam_pos(cam_coord.x, cam_coord.y, cam_coord.z);
+        int dist = (int)floor(m_boxDimensions / m_boxDensity * 0.5);
+        int cam_x = (int)floor(cam_pos.X / m_boxDensity);
+        int cam_y = (int)floor(cam_pos.Y / m_boxDensity);
```

Comment:
Maybe we add an assert for `(abs(m_boxDensity) >= FLT_EPSILON)` for robustness sake ?